src/pkg/migrate/migrate.go

```diff
@@ -0,0 +1,245 @@
+
+	term.Info("Generating compose file...")
+
+	composeFileContents, err := generateComposeFile(ctx, fabric, defangv1.SourcePlatform_SOURCE_PLATFORM_HEROKU, sourceApp, sanitizedApplicationInfo)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate compose file from Heroku info: %w", err)
+	}
+
+	return composeFileContents, nil
+}
+
+func sanitizeHerokuApplicationInfo(info HerokuApplicationInfo) (interface{}, error) {
+	for key, value := range info.ConfigVars {
+		// Redact sensitive information in config vars
+		isSecret, err := compose.IsSecret(value)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check if config var %q is a secret: %w", key, err)
+		}
+		if isSecret {
+			info.ConfigVars[key] = "REDACTED"
+		}
+	}
+
+	return info, nil
+}
+
+func generateComposeFile(ctx context.Context, fabric client.FabricClient, platform defangv1.SourcePlatform, projectName string, data any) (string, error) {
+	var err error
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal data to json: %w", err)
+	}
+
+	var resp *defangv1.GenerateComposeResponse
+	var previousError string
+	for range [3]int{} {
+		resp, err = fabric.GenerateCompose(ctx, &defangv1.GenerateComposeRequest{
+			Platform:      platform,
+			Data:          dataJSON,
+			PreviousError: previousError,
+		})
+		if err != nil {
+			return "", err
+		}
+
+		responseStr := string(resp.Compose)
+		term.Debugf("Received compose response: %+v", responseStr)
+
+		// assume the response is markdown,
+		// extract the contents of the first code block if there is one
+		composeContent, err := extractFirstCodeBlock(responseStr)
+		if err != nil {
+			if errors.Is(err, errNoCodeBlock) {
+				// If no code block found, use the entire response
+				composeContent = responseStr
+			} else {
+				previousError = err.Error()
+				term.Debugf("Failed to extract code block: %v. Retrying...", err)
+				continue
+			}
+		}
+
+		// Cleanup the compose content
+		composeContent, err = cleanupComposeFile(composeContent)
+		if err != nil {
+			previousError = err.Error()
+			continue
+		}
+
+		// Attempt to load the compose content
+		_, err = compose.LoadFromContent(ctx, []byte(composeContent), projectName)
+		if err != nil {
+			previousError = err.Error()
+			term.Debugf("Invalid compose file received: %v. Retrying...", err)
+			continue
+		}
+
+		// If we reach here, the compose content is valid
+		return composeContent, nil
+	}
+
+	if err != nil {
+		return "", fmt.Errorf("failed to generate compose file after retries: %w", err)
+	}
+
+	// This should not be reached, but just in case
+	return "", errors.New("unexpected error: no valid compose file generated")
+}
+
+func cleanupComposeFile(composeContent string) (string, error) {
+	// parse as yaml and remove `version` property if present
+	var composeYAML map[string]interface{}
+	if err := yaml.Unmarshal([]byte(composeContent), &composeYAML); err != nil {
```

Comment:
There's a way to "parse" YAML while keeping the comments around, see https://stackoverflow.com/questions/62062329/how-to-parse-general-yaml-in-golang-with-comments-preserved